File: src/main/java/io/github/svegon/capi/util/ItemUtil.java
package io.github.svegon.capi.util;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.Inventories;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.screen.ScreenHandler;
import net.minecraft.screen.slot.CraftingResultSlot;
import net.minecraft.screen.slot.FurnaceOutputSlot;
import net.minecraft.screen.slot.Slot;
import net.minecraft.screen.slot.TradeOutputSlot;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.RaycastContext;
import net.minecraft.world.World;
import org.jetbrains.annotations.NotNull;

import java.util.AbstractList;
import java.util.Collection;
import java.util.List;
import java.util.function.Predicate;
import java.util.function.ToIntFunction;

public final class ItemUtil {
    private ItemUtil() {
        throw new UnsupportedOperationException();
    }

    public static final Predicate<Slot> PERMANENT_SLOT = (slot) -> slot.getClass() == Slot.class;
    public static final Predicate<Slot> ITEM_CONTAINER_SLOT = (slot) -> !(slot instanceof FurnaceOutputSlot
            || slot instanceof CraftingResultSlot || slot instanceof TradeOutputSlot);

    public static int getSlotMatching(@NotNull final Inventory inv,
                                      @NotNull final Predicate<ItemStack> itemPredicate) {
        for (int i = 0; i < inv.size(); i++) {
            if (itemPredicate.test(inv.getStack(i)))
                return i;
        }

        return -1;
    }

    public static int getSlotWithItem(@NotNull final Inventory inv, @NotNull final Item item) {
        return getSlotMatching(inv, (stack) -> !stack.isEmpty() && stack.isOf(item));
    }

    public static int getSlotMatching(@NotNull final ScreenHandler handler, @NotNull final Predicate<Slot>
            slotPredicate, @NotNull final Predicate<ItemStack> itemPredicate) {
        for (int i = 0; i < handler.slots.size(); i++) {
            Slot slot = handler.slots.get(i);

            if (slotPredicate.test(slot)) {
                ItemStack stack = slot.getStack();

                if (itemPredicate.test(stack)) {
                    return i;
                }
            }
        }

        return -1;
    }

    public static int getSlotMatching(@NotNull final ScreenHandler handler,
                                      @NotNull final Predicate<ItemStack> itemPredicate) {
        return getSlotMatching(handler, ITEM_CONTAINER_SLOT, itemPredicate);
    }

    public static int getSlotMatching(@NotNull final ScreenHandler handler, @NotNull final Predicate<Slot>
            slotPredicate, final @NotNull Item item) {
        return getSlotMatching(handler, slotPredicate, (stack) -> !stack.isEmpty() && stack.isOf(item));
    }

    public static int getSlotWithItem(@NotNull final ScreenHandler handler, final @NotNull Item item) {
        return getSlotMatching(handler, ITEM_CONTAINER_SLOT, item);
    }

    public static int getSlotWithAny(@NotNull final Inventory inv, @NotNull final Collection<Item> items) {
        return getSlotMatching(inv, (stack) -> !stack.isEmpty() && items.contains(stack.getItem()));
    }

    public static int getSlotWithAny(@NotNull final ScreenHandler handler, @NotNull final Collection<Item> items) {
        return getSlotMatching(handler, (stack) -> !stack.isEmpty() && items.contains(stack.getItem()));
    }

    public static int getEmptySlot(@NotNull final ScreenHandler handler) {
        return getSlotMatching(handler, ItemStack::isEmpty);
    }

    /**
     * taken from Item.raycast
     *
     * @param world
     * @param player
     * @param fluidHandling
     * @return
     */
    public static BlockHitResult raycast(@NotNull final World world, @NotNull final PlayerEntity player,
                                         @NotNull final RaycastContext.FluidHandling fluidHandling) {
        float f = player.getPitch();
        float g = player.getYaw();
        Vec3d vec3d = player.getCameraPosVec(1.0F);
        float h = MathHelper.cos(-g * 0.017453292F - 3.1415927F);
        float i = MathHelper.sin(-g * 0.017453292F - 3.1415927F);
        float j = -MathHelper.cos(-f * 0.017453292F);
        float k = MathHelper.sin(-f * 0.017453292F);
        float l = i * j;
        float n = h * j;
        Vec3d vec3d2 = vec3d.add((double)l * 5.0D, (double)k * 5.0D, (double)n * 5.0D);
        return world.raycast(new RaycastContext(vec3d, vec3d2,
                RaycastContext.ShapeType.OUTLINE, fluidHandling, player));
    }

    public static int countItems(@NotNull final Inventory inv, @NotNull final ToIntFunction<ItemStack> stackToCount) {
        int count = 0;

        for (int slot = 0; slot < inv.size(); ++slot) {
            count += stackToCount.applyAsInt(inv.getStack(slot));
        }

        return count;
    }

    public static int countItems(@NotNull final Inventory inv, @NotNull final Item item) {
        return countItems(inv, (stack) -> !stack.isEmpty() && stack.isOf(item) ? stack.getCount() : 0);
    }

    public static List<ItemStack> asList(@NotNull final Inventory inv) {
        return new AbstractList<>() {
            @Override
            public ItemStack get(int index) {
                return inv.getStack(index);
            }

            @Override
            public ItemStack set(int index, ItemStack element) {
                ItemStack result = get(index);
                inv.setStack(index, element);
                return result;
            }

            @Override
            public ItemStack remove(int index) {
                return Inventories.removeStack(this, index);
            }

            @Override
            public int size() {
                return inv.size();
            }

            @Override
            public boolean isEmpty() {
                return inv.isEmpty();
            }
        };
    }

    public static Inventory asInv(@NotNull final ScreenHandler screenHandler) {
        return new Inventory() {
            @Override
            public int size() {
                return screenHandler.slots.size();
            }

            @Override
            public boolean isEmpty() {
                return screenHandler.slots.isEmpty();
            }

            @Override
            public ItemStack getStack(int slot) {
                return screenHandler.getSlot(slot).getStack();
            }

            @Override
            public ItemStack removeStack(int slot, int amount) {
                ItemStack stack = getStack(slot);
                stack.decrement(amount);
                return stack;
            }

            @Override
            public ItemStack removeStack(int slot) {
                if (slot < 0 || slot >= size()) {
                    return ItemStack.EMPTY;
                }

                Slot s = screenHandler.getSlot(slot);
                ItemStack result = s.getStack();
                s.setStack(ItemStack.EMPTY);
                return result;
            }

            @Override
            public void setStack(int slot, ItemStack stack) {
                screenHandler.getSlot(slot).setStack(stack);
            }

            @Override
            public void markDirty() {

            }

            @Override
            public boolean canPlayerUse(PlayerEntity player) {
                return screenHandler.canUse(player);
            }

            @Override
            public void clear() {
                screenHandler.slots.forEach(slot -> slot.setStack(ItemStack.EMPTY));
            }
        };
    }

    public static List<ItemStack> asList(@NotNull final ScreenHandler screenHandler) {
        return asList(asInv(screenHandler));
    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/util/ItemUtil.kt
package io.github.svegon.mclientapi.util

import net.minecraft.world.Container
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.FurnaceResultSlot
import net.minecraft.world.inventory.MerchantResultSlot
import net.minecraft.world.inventory.ResultSlot
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.ClipContext
import net.minecraft.world.level.Level
import net.minecraft.world.phys.BlockHitResult
import java.util.function.Consumer
import java.util.function.Predicate
import java.util.function.ToIntFunction

object ItemUtil {
    val PERMANENT_SLOT = { slot: Slot -> slot::class == Slot::class }
    val OUTPUT_SLOT = { slot: Slot ->
        !(slot is FurnaceResultSlot || slot is ResultSlot || slot is MerchantResultSlot)
    }
    val ITEM_CONTAINER_SLOT = { slot: Slot -> !OUTPUT_SLOT(slot)}

    fun Container.getSlotMatching(
        itemPredicate: (ItemStack) -> Boolean,
    ): Int {
        for (i in 0 until containerSize) {
            if (itemPredicate(getItem(i))) return i
        }

        return -1
    }

    fun Container.getSlotWithItem(item: Item): Int {
        return getSlotMatching { stack: ItemStack -> stack.`is`(item) }
    }

    fun AbstractContainerMenu.getSlotMatching(
        slotPredicate: Predicate<Slot>,
        itemPredicate: Predicate<ItemStack>,
    ): Int {
        for (i in slots.indices) {
            val slot = slots[i]

            if (slotPredicate.test(slot)) {
                val stack = slot.item

                if (itemPredicate.test(stack)) {
                    return i
                }
            }
        }

        return -1
    }

    fun AbstractContainerMenu.getSlotMatching(
        itemPredicate: Predicate<ItemStack>,
    ): Int {
        return getSlotMatching(ITEM_CONTAINER_SLOT, itemPredicate)
    }

    fun AbstractContainerMenu.getSlotMatching(slotPredicate: Predicate<Slot>, item: Item): Int {
        return getSlotMatching(slotPredicate) { stack: ItemStack -> stack.`is`(item) }
    }

    fun AbstractContainerMenu.getSlotWithItem(item: Item): Int {
        return getSlotMatching(ITEM_CONTAINER_SLOT, item)
    }

    fun Container.getSlotWithAny(items: Collection<Item>): Int {
        return getSlotMatching { stack: ItemStack -> items.contains(stack.item) }
    }

    fun AbstractContainerMenu.getSlotWithAny(items: Collection<Item>): Int {
        return getSlotMatching { stack: ItemStack -> items.contains(stack.item) }
    }

    fun AbstractContainerMenu.getEmptySlot(): Int {
        return getSlotMatching { obj: ItemStack -> obj.isEmpty }
    }

    fun Container.countItems(stackToCount: ToIntFunction<ItemStack>): Int {
        var count = 0

        for (slot in 0 until containerSize) {
            count += stackToCount.applyAsInt(getItem(slot))
        }

        return count
    }

    fun Container.countItems(item: Item): Int {
        return countItems { stack: ItemStack -> if (!stack.isEmpty && stack.`is`(item)) stack.count else 0 }
    }

    fun Inventory.findHotbarSlotMatching(itemPredicate: (ItemStack) -> Boolean): Int {
        for (i in 0..<Inventory.getSelectionSize()) {
            if (itemPredicate(getItem(i))) {
                return i
            }
        }

        return -1
    }

    fun Inventory.findHotbarSlotWith(item: Item): Int {
        return findHotbarSlotMatching { stack: ItemStack -> stack.`is`(item) }
    }

    fun Inventory.findHotbarSlotWith(items: Collection<Item>): Int {
        return findHotbarSlot { item: Item -> items.contains(item) }
    }

    fun Inventory.findHotbarSlot(itemPredicate: (Item) -> Boolean): Int {
        return findHotbarSlotMatching { stack: ItemStack -> itemPredicate(stack.item) }
    }

    fun Inventory.getEmptyHotbarSlot(): Int {
        return findHotbarSlotMatching { stack: ItemStack -> stack.isEmpty }
    }

    /**
     * taken from Item.raycast
     *
     * @param level
     * @param player
     * @param fluidHandling
     * @return
     */
    fun raycast(
        level: Level, player: Player,
        fluidHandling: ClipContext.Fluid,
    ): BlockHitResult {
        val vec3d = player.eyePosition
        val vec3d2 = vec3d.add(player.calculateViewVector(player.xRot, player.yRot)
            .scale(player.blockInteractionRange()))
        return level.clip(ClipContext(vec3d, vec3d2, ClipContext.Block.OUTLINE, fluidHandling, player))
    }

    fun AbstractContainerMenu.asInv(): Container {
        return object : Container {
            override fun getContainerSize(): Int {
                return slots.size
            }

            override fun isEmpty(): Boolean {
                return slots.isEmpty()
            }

            override fun getItem(slot: Int): ItemStack {
                return getSlot(slot)!!.get();
            }

            override fun removeItem(slot: Int, count: Int): ItemStack {
                val s = getSlot(slot) ?: return ItemStack.EMPTY

                val result = s.get();
                s.set(ItemStack.EMPTY)
                return result
            }

            override fun removeItemNoUpdate(slot: Int): ItemStack {
                val s = getSlot(slot) ?: return ItemStack.EMPTY
                val result = s.get();
                s.set(ItemStack.EMPTY)
                return result
            }

            override fun setItem(slot: Int, itemStack: ItemStack) {
                val s = getSlot(slot) ?: return
                s.set(itemStack)
            }

            override fun setChanged() {}

            override fun stillValid(player: Player): Boolean {
                return true
            }

            override fun clearContent() {
                slots.forEach(Consumer { slot: Slot -> slot.set(ItemStack.EMPTY) })
            }
        }
    }
}
